README.md
WooCommerce REST API PHP Client Library
=======================================

## About

A PHP wrapper for the WooCommerce REST API. Easily interact with the WooCommerce REST API using this library.

Feedback and bug reports are appreciated.

## Requirements

PHP 5.2.x
cURL
WooCommerce 2.5 at least on the store

## Getting started

Generate API credentials (Consumer Key & Consumer Secret) under WP Admin > Your Profile.

## Setup the library

```php
require_once( 'lib/woocommerce-api.php' );

$options = array(
	'ssl_verify'      => false,
);

try {

	$client = new WC_API_Client( 'http://your-store-url.com', $consumer_key, $consumer_secret, $options );

} catch ( WC_API_Client_Exception $e ) {

	echo $e->getMessage() . PHP_EOL;
	echo $e->getCode() . PHP_EOL;

	if ( $e instanceof WC_API_Client_HTTP_Exception ) {

		print_r( $e->get_request() );
		print_r( $e->get_response() );
	}
}
```

### Options

* `debug` (default `false`) - set to `true` to add request/response information to the returned data. This is particularly useful for troubleshooting errors.

* `return_as_array` (default `false`) - all methods return data as a `stdClass` by default, but you can set this option to `true` to return data as an associative array instead.

* `validate_url` (default `false`) - set this to `true` to verify that the URL provided has a valid, parseable WC API index, and optionally force SSL when supported.

* `timeout` (default `30`) - set this to control the HTTP timeout for requests.

* `ssl_verify` (default `true`) - set this to `false` if you don't want to perform SSL peer verification for every request.


### Error handling
Exceptions are thrown when errors are encountered, most will be instances of `WC_API_Client_HTTP_Exception` which has two additional methods, `get_request()` and `get_response()` -- these return the request and response objects to help with debugging.


## Methods

### Index

* `$client->index->get()` - get the API index

### Orders

* `$client->orders->get()` - get a list of orders
* `$client->orders->get( null, array( 'status' => 'completed' ) )` - get a list of completed orders
* `$client->orders->get( $order_id )` - get a single order


## Credit

Copyright (c) 2013-2014 - [Gerhard Potgieter](http://gerhardpotgieter.com/), [Max Rice](http://maxrice.com) and other contributors

## License
Released under the [GPL3 license](http://www.gnu.org/licenses/gpl-3.0.html)


CHANGELOG
## Changelog
**version 3.0.0 - 2016-01-23**
- Add support for v3 API and creating / updating product categories
- Stop throwing exception on sku not found

**version 2.0.1 - 2015-07-13**
- Fix composer configuration

**version 2.0.0 - 2015-05-05**
- Add support for v2 API (v1 is no longer supported)
- Greatly improved error handling and logging
- Add store URL validation

**version 0.3.1 - 2014-05-02**

- Fix parameter normalization issue with WC 2.1.7+

**version 0.3 - 2014-02-20**

- Add HTTP error messages on failed cURL calls

**version 0.2 - 2014-01-22**

- Add support for filters/params to endpoint functions
- Add new top sellers report endpoint function
- Add function to call custom endpoints

**version 0.1 - 2013-12-10**

- Initial release


lib/woocommerce-api/class-wc-api-client-http-request.php
<?php
/**
 * WC API Client HTTP Request
 *
 * Handles HTTP communication to the WC API via cURL
 *
 * @since 2.0
 */
class WC_API_Client_HTTP_Request {


	/** @var resource cURL handle */
	protected $ch;

	/** @var stdClass request data */
	protected $request;

	/** @var bool true to decode JSON as array */
	protected $json_decode_as_array;

	/** @var bool true to include cURL log, HTTP request/response object in result */
	protected $debug;

	/** @var stdClass response data */
	protected $response;

	/** @var string raw HTTP response headers */
	protected $curl_headers;


	/**
	 * Setup HTTP request
	 *
	 * @since 2.0
	 * @param array $args request args
	 */
	public function __construct( $args ) {

		$this->request = new stdClass();

		$this->request->headers = array(
			'Accept: application/json',
			'Content-Type: application/json',
			'User-Agent: WooCommerce API Client-PHP/' . WC_API_Client::VERSION,
		);

		// GET, POST, PUT, DELETE, etc.
		$this->request->method = $args['method'];

		// trailing slashes tend to cause OAuth authentication issues, so strip them
		$this->request->url = rtrim( $args['url'], '/' );

		$this->request->params = array();
		$this->request->data = $args['data'];

		// JSON output format
		$this->json_decode_as_array = ( 'array' === $args['options']['json_decode'] );

		// debug mode?
		$this->debug = (bool) $args['options']['debug'];

		// optional cURL opts
		$timeout = (int) $args['options']['timeout'];
		$ssl_verify = (bool) $args['options']['ssl_verify'];

		$this->ch = curl_init();

		// default cURL opts
		curl_setopt( $this->ch, CURLOPT_SSL_VERIFYPEER, $ssl_verify );
		curl_setopt( $this->ch, CURLOPT_SSL_VERIFYHOST, $ssl_verify );
		curl_setopt( $this->ch, CURLOPT_CONNECTTIMEOUT, $timeout );
		curl_setopt( $this->ch, CURLOPT_TIMEOUT, (int) $timeout );
		curl_setopt( $this->ch, CURLOPT_RETURNTRANSFER, true );

		// set request headers
		curl_setopt( $this->ch, CURLOPT_HTTPHEADER, $this->request->headers );

		// save response headers
		curl_setopt( $this->ch, CURLOPT_HEADERFUNCTION, array( $this, 'curl_stream_headers' ) );

		// set request method and data
		switch ( $this->request->method ) {

			case 'GET':
				$this->request->body = null;
				$this->request->params = (array) $this->request->data;
			break;

			case 'PUT':
				$this->request->body = json_encode( $this->request->data );
				curl_setopt( $this->ch, CURLOPT_CUSTOMREQUEST, 'PUT' );
				curl_setopt( $this->ch, CURLOPT_POSTFIELDS, $this->request->body );
			break;

			case 'POST':
				$this->request->body = json_encode( $this->request->data );
				curl_setopt( $this->ch, CURLOPT_POST, true );
				curl_setopt( $this->ch, CURLOPT_POSTFIELDS, $this->request->body );
			break;

			case 'DELETE':
				$this->request->body = null;
				$this->request->params = (array) $this->request->data;
				curl_setopt( $this->ch, CURLOPT_CUSTOMREQUEST, 'DELETE' );
				break;
		}

		// auth
		$this->request->url = $this->get_url_with_auth( $args['consumer_key'], $args['consumer_secret'] );

		// set request url
		curl_setopt( $this->ch, CURLOPT_URL, $this->request->url );
	}


	/**
	 * Set authentication parameters for the request
	 *
	 * @since 2.0
	 * @param string $consumer_key API consumer key
	 * @param string $consumer_secret API consumer secret
	 * @return string request URL with authentication parameters added
	 */
	protected function get_url_with_auth( $consumer_key, $consumer_secret ) {

		$auth = new WC_API_Client_Authentication( $this->request->url, $consumer_key, $consumer_secret );

		if ( $auth->is_ssl() ) {

			// query string authentication over SSL works with all servers, whereas HTTP basic auth fails in certain cases
			// see https://github.com/kloon/WooCommerce-REST-API-Client-Library/issues/29
			$this->request->params = array_merge( $this->request->params, array(
				'consumer_key'    => $auth->get_consumer_key(),
				'consumer_secret' => $auth->get_consumer_secret(),
			) );

		} else {

			$this->request->params = array_merge( $this->request->params, $auth->get_oauth_params( $this->request->params, $this->request->method ) );
		}

		// build url
		return $this->request->url . ( ! empty( $this->request->params ) ? '?' . http_build_query( $this->request->params ) : '' );
	}


	/**
	 * Send the request
	 *
	 * @since 2.0
	 * @return object|array result
	 * @throws WC_API_Client_HTTP_Exception invalid decoded JSON or any non-HTTP 200/201 response
	 */
	public function dispatch() {

		$this->response = new stdClass();

		// blank headers
		$this->curl_headers = '';

		$start_time = microtime( true );

		// send request + save raw response body
		$this->response->body = curl_exec( $this->ch );

		// request duration
		$this->request->duration = round( microtime( true ) - $start_time, 5 );

		// response code
		$this->response->code = curl_getinfo( $this->ch, CURLINFO_HTTP_CODE );

		// response headers
		$this->response->headers = $this->get_response_headers();

		curl_close( $this->ch );

		$parsed_response = $this->get_parsed_response( $this->response->body );

		// check for invalid JSON
		if ( null === $parsed_response ) {

			throw new WC_API_Client_HTTP_Exception( sprintf( 'Invalid JSON returned for %s.', $this->request->url ), $this->response->code, $this->request, $this->response );
		}

		// any non-200/201/202 response code indicates an error
		if ( ! in_array( $this->response->code, array( '200', '201', '202' ) ) ) {

			// error message/code is nested sometimes
			if ( $this->json_decode_as_array ) {

				list( $error_message, $error_code ) = is_array( $parsed_response['errors'] ) ? array(
					$parsed_response['errors'][0]['message'],
					$parsed_response['errors'][0]['code']
				) : array(
					$parsed_response['errors']['message'],
					$parsed_response['errors']['code']
				);

			} else {

				list( $error_message, $error_code ) = is_array( $parsed_response->errors ) ? array(
					$parsed_response->errors[0]->message,
					$parsed_response->errors[0]->code
				) : array(
					$parsed_response->errors->message,
					$parsed_response->errors->code
				);
			}
			
			// Stop throwing exceptions when sku isn't found
			if($error_code!="woocommerce_api_invalid_product_sku"){
				throw new WC_API_Client_HTTP_Exception( sprintf( 'Error: %s [%s]', $error_message, $error_code ), $this->response->code, $this->request, $this->response);
			}
		}

		return $this->build_result( $parsed_response );
	}


	/**
	 * JSON decode the response body after stripping any invalid leading or
	 * trailing characters.
	 *
	 * Plugins (looking at you WP Super Cache) or themes
	 * can add output to the returned JSON which breaks decoding.
	 *
	 * @since 2.0
	 * @param string $raw_body raw response body
	 * @return object|array JSON decoded response body
	 */
	protected function get_parsed_response( $raw_body ) {

		$json_start = strpos( $raw_body, '{' );
		$json_end = strrpos( $raw_body, '}' ) + 1; // inclusive

		$json = substr( $raw_body, $json_start, ( $json_end - $json_start ) );

		return json_decode( $json, $this->json_decode_as_array );
	}


	/**
	 * Build the result object/array
	 *
	 * @since 2.0.0
	 * @param object|array JSON decoded result
	 * @return object|array in format:
	 * {
	 *  <result data>
	 *  'http' =>
	 *   'request' => stdClass(
	 *     'url' => request URL
	 *     'method' => request method
	 *     'body' => JSON encoded request body entity
	 *     'headers' => array of request headers
	 *     'duration' => request duration, in seconds
	 *     'params' => optional raw params
	 *     'data' => optional raw request data
	 *     'duration' =>
	 *    )
	 *   'response' => stdClass(
	 *     'body' => raw response body
	 *     'code' => HTTP response code
	 *     'headers' => HTTP response headers in assoc array
	 *   )
	 * }
	 */
	protected function build_result( $parsed_response ) {

		// add cURL log, HTTP request/response object
		if ( $this->debug ) {

			if ( $this->json_decode_as_array ) {

				$parsed_response['http'] = array(
					'request'  => json_decode( json_encode( $this->request ), true ),
					'response' => json_decode( json_encode( $this->response ), true ),
				);

			} else {

				$parsed_response->http = new stdClass();
				$parsed_response->http->request = $this->request;
				$parsed_response->http->response = $this->response;
			}
		}

		return $parsed_response;
	}


	/**
	 * Save the cURL response headers for later processing
	 *
	 * @since 2.0
	 * @see WP_Http_Curl::stream_headers()
	 * @param object $_ the cURL resource handle (unused)
	 * @param string $headers the current response headers
	 * @return int the size of the processed headers
	 */
	public function curl_stream_headers( $_, $headers ) {

		$this->curl_headers .= $headers;
		return strlen( $headers );
	}


	/**
	 * Parse the raw response headers into an assoc array in format:
	 * {
	 *   'Header-Key' => header value
	 *   'Duplicate-Key' => array(
	 *     0 => value 1
	 *     1 => value 2
	 *   )
	 * }
	 *
	 * @since 2.0
	 * @see WP_HTTP::processHeaders
	 * @return array
	 */
	protected function get_response_headers() {

		// get the raw headers
		$raw_headers = preg_replace('/\n[ \t]/', ' ', str_replace( "\r\n", "\n", $this->curl_headers ) );

		// spit them
		$raw_headers = array_filter( explode( "\n", $raw_headers ), 'strlen' );

		$headers = array();

		// parse into assoc array
		foreach ( $raw_headers as $header ) {

			// skip response codes (appears as HTTP/1.1 200 OK or HTTP/1.1 100 Continue)
			if ( 'HTTP/' === substr( $header, 0, 5 ) ) {
				continue;
			}

			list( $key, $value ) = explode( ':', $header, 2 );

			if ( isset( $headers[ $key ] ) ) {

				// ensure duplicate headers aren't overwritten
				$headers[ $key ] = array( $headers[ $key ] );
				$headers[ $key ][] = $value;

			} else {
				$headers[ $key ] = $value;
			}

		}

		return $headers;
	}

}


lib/woocommerce-api/resources/abstract-wc-api-client-resource.php
<?php
/**
 * WC API Client Resource class
 *
 * @since 2.0
 */
abstract class WC_API_Client_Resource {


	/** @var string resource endpoint */
	protected $endpoint;

	/** @var string JSON object namespace */
	protected $object_namespace;

	/** @var WC_API_Client class instance */
	protected $client;

	/** @var string request method, e.g. GET */
	protected $request_method;

	/** @var string request path, e.g. orders/123 */
	protected $request_path;

	/** @var array request params, e.g. { 'status' => 'processing' } */
	protected $request_params;

	/** @var array request body data, only used for PUT/POST requests */
	protected $request_body;


	/**
	 * Set the endpoint and client
	 *
	 * @since 2.0
	 * @param string $endpoint top-level endpoint, e.g. `orders`
	 * @param string $object_namespace the JSON object namespace for this resource, e.g. `order`
	 * @param WC_API_Client $client class instance
	 */
	public function __construct( $endpoint, $object_namespace, $client ) {

		$this->endpoint         = $endpoint;
		$this->object_namespace = $object_namespace;
		$this->client           = $client;
	}


	/**
	 * Set the arguments for the request, required:
	 *
	 * `method`
	 * `path`
	 *
	 * optional:
	 *
	 * `params`
	 * `body`
	 *
	 * @since 2.0
	 * @param array $args
	 */
	protected function set_request_args( $args ) {

		$this->request_method = $args['method'];
		$this->request_path   = isset( $args['path'] ) ? $args['path'] : null;
		$this->request_params = isset( $args['params'] ) ? $args['params'] : null;
		$this->request_body   = isset( $args['body'] ) ? $args['body'] : null;

		// set the top-level JSON object namespace if not already set, this is mainly
		// a convenience for client code so creating/updating resources doesn't need a
		// nested array like array( 'order_note' => array( 'note' => 'foo' ) ) and can instead
		// use array( 'note' => 'foo' ) ʘ‿ʘ
		// prevent 'product' namespace from being added to product_category call
		if ( $this->request_body && ! isset( $args['body'][ $this->object_namespace ] ) && sizeof( $args['body']['product_category'] )==0 ) {

			$this->request_body = array(
				$this->object_namespace => $this->request_body,
			);
		}

		// convert bool true to string 'true', required for DELETE endpoints
		if ( isset( $this->request_params['force'] ) && $this->request_params['force'] ) {
			$this->request_params['force'] = 'true';
		}
	}


	/**
	 * Return the full endpoint path, e.g. given an endpoint of `orders` and a
	 * path of `123`, return `orders/123`. For nested resources, this will return
	 * the full path as well
	 *
	 * @since 2.0
	 * @return string
	 */
	protected function get_endpoint_path() {

		return empty( $this->request_path ) ? $this->endpoint : $this->endpoint . '/' . implode( '/', (array) $this->request_path );
	}


	/**
	 * Return the request data, either query parameters (for GET/DELETE requests)
	 * or the request body (for PUT/POST requests)
	 *
	 * @since 2.0
	 * @return array
	 */
	protected function get_request_data() {

		return ( 'GET' === $this->request_method || 'DELETE' === $this->request_method ) ? $this->request_params : $this->request_body;
	}


	/**
	 * Perform the request and return the response
	 *
	 * @since 2.0
	 * @return array|object
	 */
	protected function do_request() {

		return $this->client->make_api_call( $this->request_method, $this->get_endpoint_path(), $this->get_request_data() );
	}


}




lib/woocommerce-api/resources/class-wc-api-client-resource-products.php
<?php
/**
 * WC API Client Products resource class
 *
 * @since 2.0
 */
class WC_API_Client_Resource_Products extends WC_API_Client_Resource {


	/**
	 * Setup the resource
	 *
	 * @since 2.0
	 * @param WC_API_Client $client class instance
	 */
	public function __construct( $client ) {

		parent::__construct( 'products', 'product', $client );
	}


	/**
	 * Get products
	 *
	 * GET /products
	 * GET /products/#{id}
	 *
	 * @since 2.0
	 * @param null|int $id product ID or null to get all products
	 * @param array $args acceptable products endpoint args, like `filter[]`
	 * @return array|object products!
	 */
	public function get( $id = null, $args = array() ) {

		$this->set_request_args( array(
			'method' => 'GET',
			'path'   => $id,
			'params' => $args,
		) );

		return $this->do_request();
	}


	/**
	 * Get product by SKU
	 *
	 * GET /products/sku/{sku}
	 *
	 * Note this will throw an exception if no products are found (404 not found)
	 *
	 * @since 2.0
	 * @param string $sku product SKU
	 * @param array $args acceptable product SKU lookup endpoint args, currently only `fields`
	 * @return array|object product!
	 */
	public function get_by_sku( $sku, $args = array() ) {

		$this->set_request_args( array(
			'method' => 'GET',
			'path'   => array( 'sku', urlencode( $sku ) ),
			'params' => $args,
		) );

		return $this->do_request();
	}


	/**
	 * Create a product
	 *
	 * POST /products
	 *
	 * @since 2.0
	 * @param array $data valid product data
	 * @return array|object your newly-created product
	 */
	public function create( $data ) {

		$this->set_request_args( array(
			'method' => 'POST',
			'body'   => $data,
		) );

		return $this->do_request();
	}


	/**
	 * Update a product
	 *
	 * PUT /product/#{id}
	 *
	 * @since 2.0
	 * @param int $id product ID
	 * @param array $data product data to update
	 * @return array|object your newly-updated product
	 */
	public function update( $id, $data ) {

		$this->set_request_args( array(
			'method' => 'PUT',
			'path'   => $id,
			'body'   => $data,
		) );

		return $this->do_request();
	}


	/**
	 * Delete a product
	 *
	 * DELETE /products/#{id}
	 *
	 * @since 2.0
	 * @param int $id product ID
	 * @param bool $force true to permanently delete the product, false to trash it
	 * @return array|object response
	 */
	public function delete( $id, $force = false ) {

		$this->set_request_args( array(
			'method' => 'DELETE',
			'path'   => $id,
			'params' => array( 'force' => $force ),
		) );

		return $this->do_request();
	}


	/**
	 * Get a count of products
	 *
	 * GET /products/count
	 *
	 * @since 2.0
	 * @param array $args acceptable product endpoint args, like `type` or `filter[]`
	 * @return array|object the count
	 */
	public function get_count( $args = array() ) {

		$this->set_request_args( array(
			'method' => 'GET',
			'path'   => 'count',
			'params' => $args,
		) );

		return $this->do_request();
	}


	/**
	 * Get product reviews
	 *
	 * GET /products/#{product_id}/reviews
	 *
	 * @since 2.0
	 * @param int $id product ID
	 * @param array $args acceptable product reviews endpoint args, currently only `fields`
	 * @return array|object product reviews!
	 */
	public function get_reviews( $id, $args = array() ) {

		$this->set_request_args( array(
			'method' => 'GET',
			'path'   => array( $id, 'reviews' ),
			'params' => $args,
		) );

		return $this->do_request();
	}


	/**
	 * Get a list of product categories or a single product category
	 *
	 * GET /products/categories
	 * GET /products/categories/{#id}
	 *
	 * @since 2.0
	 * @param int $id category ID or null to get all product categories
	 * @param array $args acceptable product categories endpoint args, currently only `fields`
	 * @return array|object product categories!
	 */
	public function get_categories( $id = null, $args = array() ) {

		$this->set_request_args( array(
			'method' => 'GET',
			'path'   => array( 'categories', $id ),
			'params' => $args,
		) );

		return $this->do_request();
	}

	/**
	 * Create a category
	 *
	 * POST /products/categories
	 *
	 * @since 2.5
	 * @param array $data valid product category
	 * @return array|object your newly-created category
	 */
	public function create_category( $data ) {

		$this->set_request_args( array(
			'method' => 'POST',
                        'path'   => array( 'categories' ),
			'body'   => $data,
		) );

		return $this->do_request();
	}
        
        
        /**
	 * Update a category
	 *
	 * POST /products/categories/{#id}
	 *
	 * @since 2.5
	 * @param array $data valid category data
	 * @return array|object your newly-created category
	 */
	public function update_category( $data, $id ) {

		$this->set_request_args( array(
			'method' => 'POST',
                        'path'   => array( 'categories', $id ),
			'body'   => $data,
		) );

		return $this->do_request();
	}

	/** Convenience methods - these do not map directly to an endpoint ********/


	/**
	 * Update the stock quantity for a product
	 *
	 * PUT /products/#{id} with stock quantity
	 *
	 * @param int $id order ID
	 * @param int|float $quantity new stock quantity
	 * @return array|object newly-updated product
	 */
	public function update_stock( $id, $quantity ) {

		$this->set_request_args( array(
			'method' => 'PUT',
			'path'   => $id,
			'body'   => array( 'stock_quantity' => $quantity ),
		) );

		return $this->do_request();
	}


}


lib/woocommerce-api/class-wc-api-client.php
<?php
/**
 * WC API Client
 *
 * @since 3.0
 */
class WC_API_Client {


	/** API client version */
	const VERSION = '3.0.0';

	/** @var string store URL, e.g. http://www.woothemes.com */
	public $store_url;

	/** @var string consumer key */
	public $consumer_key;

	/** @var string consumer secret */
	public $consumer_secret;

	/** @var string API URL, e.g. http://www.woothemes.com/wc-api/v2 */
	public $api_url;

	/** @var bool true if debug is enabled */
	public $debug = false;

	/** @var bool return the response data as an array, defaults to object */
	public $return_as_array = false;

	/** @var bool perform validation on the API URL */
	public $validate_url = false;

	/** @var int HTTP request timeout */
	public $timeout = 30;

	/** @var bool true to perform SSL peer verification */
	public $ssl_verify = true;

	/** Resources */

	/** @var WC_API_Client_Resource_Coupons instance */
	public $coupons;

	/** @var WC_API_Client_Resource_Custom instance */
	public $custom;

	/** @var WC_API_Client_Resource_Customers instance */
	public $customers;

	/** @var WC_API_Client_Resource_Index instance */
	public $index;

	/** @var WC_API_Client_Resource_Orders instance */
	public $orders;

	/** @var WC_API_Client_Resource_Order_Notes instance */
	public $order_notes;

	/** @var WC_API_Client_Resource_Order_Refunds instance */
	public $order_refunds;

	/** @var WC_API_Client_Resource_Products instance */
	public $products;

	/** @var WC_API_Client_Resource_Reports instance */
	public $reports;

	/** @var WC_API_Client_Resource_Webhooks instance */
	public $webhooks;


	/**
	 * Setup the client
	 *
	 * @since 2.0
	 * @param string $store_url store URL, e.g. http://www.woothemes.com
	 * @param string $consumer_key
	 * @param string $consumer_secret
	 * @param array $options client options
	 */
	public function __construct( $store_url, $consumer_key, $consumer_secret, $options = array() ) {

		// required functions
		if ( ! extension_loaded( 'curl' ) ) {
			throw new Exception( 'WooCommerce REST API client requires the cURL PHP extension.' );
		}

		if ( ! extension_loaded( 'json' ) ) {
			throw new Exception( 'WooCommerce REST API client needs the JSON extension.' );
		}

		// set required info
		$this->store_url = $store_url;
		$this->consumer_key = $consumer_key;
		$this->consumer_secret = $consumer_secret;

		// load each API resource
		$this->init_resources();

		// build API url from store URL
		$this->build_api_url();

		// set options
		$this->parse_options( $options );

		if ( $this->validate_url ) {
			$this->validate_api_url();
		}
	}


	/**
	 * Load each API resource
	 *
	 * @since 2.0
	 */
	public function init_resources() {

		$resources = array(
			'WC_API_Client_Resource_Coupons'       => 'coupons',
			'WC_API_Client_Resource_Custom'        => 'custom',
			'WC_API_Client_Resource_Customers'     => 'customers',
			'WC_API_Client_Resource_Index'         => 'index',
			'WC_API_Client_Resource_Orders'        => 'orders',
			'WC_API_Client_Resource_Order_Notes'   => 'order_notes',
			'WC_API_Client_Resource_Order_Refunds' => 'order_refunds',
			'WC_API_Client_Resource_Products'      => 'products',
			'WC_API_Client_Resource_Reports'       => 'reports',
			'WC_API_Client_Resource_Webhooks'      => 'webhooks',
		);

		foreach ( $resources as $resource_class => $resource_method ) {

			if ( class_exists( $resource_class ) ) {
				$this->$resource_method = new $resource_class( $this );
			}
		}
	}


	/**
	 * Build the correct API URL given the store URL provided
	 *
	 * @since 2.0
	 */
	public function build_api_url() {

		$url = parse_url( $this->store_url );

		// default to http if not provided
		$scheme = isset( $url['scheme'] ) ? $url['scheme'] : 'http';

		// set host
		$host = $url['host'];

		// add port to host if provided
		$host .= isset( $url['port'] ) ? ':' . $url['port'] : '';

		// set path and strip any trailing slashes
		$path = isset( $url['path'] ) ? rtrim( $url['path'], '/' ) : '';

		// add WC API path
		$path .= '/wc-api/v3/';

		// build URL
		$this->api_url = "{$scheme}://{$host}{$path}";
	}


	/**
	 * Parse client options, current available options are:
	 *
	 * `debug` - true to include cURL log, HTTP request object, and HTTP response object in result
	 * `return_as_array` - true to return the response as an associative array instead of object
	 * `validate_url` - true to validate the API URL is correct before making API calls
	 *
	 *
	 * to implement:
	 * `timeout` - HTTP request timeout
	 * ... additional HTTP options for handling proxy servers, etc.
	 *
	 * @since 2.0
	 * @param array $options
	 */
	public function parse_options( $options ) {

		$valid_options = array(
			'debug',
			'return_as_array',
			'validate_url',
			'timeout',
			'ssl_verify',
		);

		foreach ( (array) $options as $opt_key => $opt_value ) {

			// backwards compat
			if ( 'verbose_mode' === $opt_key ) {
				$opt_key = 'debug';
			}

			if ( ! in_array( $opt_key, $valid_options ) ) {
				continue;
			}

			$this->$opt_key = $opt_value;
		}
	}


	/**
	 * Validate the API URL by checking if the API index exists and is parseable,
	 * as well as forcing SSL for stores that allow it
	 *
	 * @since 2.0
	 * @throws WC_API_Client_Exception
	 */
	public function validate_api_url() {

		$index = @file_get_contents( $this->api_url );

		// check for HTTP 404 response (file_get_contents() returns false when encountering 404)
		// this usually means:
		// 1) the store URL is not correct (missing sub-directory path, etc)
		// 2) pretty permalinks are disabled
		if ( false === $index ) {
			throw new WC_API_Client_Exception( sprintf( 'Invalid URL, no WC API found at %s -- ensure your store URL is correct and pretty permalinks are enabled.', $this->api_url ), 404 );
		}

		// older versions of WC (2.0 and under) will simply return a "1"
		if ( '1' === $index ) {
			throw new WC_API_Client_Exception( sprintf( 'Please upgrade the WooCommerce version on %s to v2.2 or greater.', $this->api_url ) );
		}

		// strip invalid leading/trailing characters from JSON
		$json_start = strpos( $index, '{' );
		$json_end = strrpos( $index, '}' ) + 1; // inclusive

		$index = json_decode( substr( $index, $json_start, ( $json_end - $json_start ) ) );

		// check for invalid JSON, an error here usually means:
		// 1) there's some garbage in the JSON output, WP Super Cache is notorious for adding an HTML comment to non-cached pages
		if ( null === $index ) {
			throw new WC_API_Client_Exception( sprintf( 'WC API found, but JSON is corrupt -- ensure the index at %s is valid JSON.', $this->api_url ) );
		}

		// check if the site URL returned is SSL, but SSL is not enabled
		if ( 'https' === parse_url( $index->store->URL, PHP_URL_SCHEME ) && ! $index->store->meta->ssl_enabled ) {

			// override the user-entered URL with the SSL version
			$this->api_url = str_replace( 'http://', 'https://', $this->api_url );
		}
	}


	/**
	 * Make the API call to specified endpoint with the given data -- also
	 * handle decoding the JSON
	 *
	 * @since 2.0
	 * @param string $method HTTP method, e.g. GET
	 * @param string $path request path, e.g. orders/123
	 * @param array $request_data either query parameters or the request body
	 * @return object|array object by default
	 * @throws WC_API_Client_Exception HTTP or authentication errors
	 */
	public function make_api_call( $method, $path, $request_data ) {

		$args = array(
			'method'          => $method,
			'url'             => $this->api_url . $path,
			'data'            => $request_data,
			'consumer_key'    => $this->consumer_key,
			'consumer_secret' => $this->consumer_secret,
			'options'         => array(
				'timeout'     => $this->timeout,
				'ssl_verify'  => $this->ssl_verify,
				'json_decode' => $this->return_as_array ? 'array' : 'object',
				'debug'       => $this->debug,
			)
		);

		$request = new WC_API_Client_HTTP_Request( $args );

		return $request->dispatch();
	}


}
